src/providers/hana/qgshanafeatureiterator.cpp

```diff
@@ -0,0 +1,437 @@
+      for ( int attrIndex : usedAttributeIndices )
+      {
+        if ( !attrs.contains( attrIndex ) )
+          attrs << attrIndex;
+      }
+    }
+  }
+
+  Q_FOREACH ( int i, attrs )
+  {
+    QString fieldname = mSource->mFields.at( i ).name();
+    if ( mFidColumn == fieldname )
+      continue;
+    mAttributesToFetch.append( i );
+  }
+
+  QString sqlFields = "";
+
+  // Add feature id column
+  if ( !mFidColumn.isEmpty() )
+    sqlFields += QgsHanaUtils::quotedIdentifier( mFidColumn ) + ",";
+
+  Q_FOREACH ( int i, mAttributesToFetch )
+  {
+    QString fieldname = mSource->mFields.at( i ).name();
+    sqlFields += QStringLiteral( "%1," ).arg( QgsHanaUtils::quotedIdentifier( fieldname ) );
+  }
+  mHasAttributes = !mAttributesToFetch.isEmpty();
+
+  // Add geometry column
+  if ( ( !( request.flags() & QgsFeatureRequest::NoGeometry )
+         || ( request.filterType() == QgsFeatureRequest::FilterExpression &&
+              request.filterExpression()->needsGeometry() ) ) && mSource->isSpatial() )
+  {
+    sqlFields += QStringLiteral( "%1" ).arg( QgsHanaUtils::quotedIdentifier( mSource->mGeometryColumn ) );
+    mHasGeometryColumn = true;
+  }
+
+  if ( sqlFields.isEmpty() )
+  {
+    sqlFields = "*";
+  }
+  else
+  {
+    if ( sqlFields.endsWith( ',' ) )
+      sqlFields.truncate( sqlFields.length() - 1 );
+  }
+
+  QString sql = QStringLiteral( "SELECT %1 FROM %2.%3" ).arg(
+                  sqlFields,
+                  QgsHanaUtils::quotedIdentifier( mSource->mSchemaName ),
+                  QgsHanaUtils::quotedIdentifier( mSource->mTableName ) );
+
+  QString sqlFilter;
+  // Set spatial filter
+  if ( !( filterRect.isNull() || filterRect.isEmpty() ) && mSource->isSpatial() && mHasGeometryColumn )
+    sqlFilter = getBBOXFilter( QgsHanaUtils::toHANAVersion( mConnection->getDatabaseVersion() ) );
+
+  if ( !mSource->mQueryWhereClause.isEmpty() )
+    sqlFilter = andWhereClauses( sqlFilter, mSource->mQueryWhereClause );
+
+  // Set fid filter
+  if ( !mFidColumn.isEmpty() )
+  {
+    if ( request.filterType() == QgsFeatureRequest::FilterFid )
+    {
+      QString inClause = QStringLiteral( " %1 = %2" ).arg(
+                           QgsHanaUtils::quotedIdentifier( mFidColumn ), FID_TO_STRING( request.filterFid() ) );
+      sqlFilter = andWhereClauses( sqlFilter, inClause );
+    }
+    else if ( request.filterType() == QgsFeatureRequest::FilterFids && !mRequest.filterFids().isEmpty() )
+    {
+      QString delim;
+      QString inClause = QStringLiteral( "%1 IN (" ).arg( QgsHanaUtils::quotedIdentifier( mFidColumn ) );
+      Q_FOREACH ( QgsFeatureId featureId, mRequest.filterFids() )
+      {
+        inClause += delim + FID_TO_STRING( featureId );
+        delim = ',';
+      }
+      inClause.append( ')' );
+
+      sqlFilter = andWhereClauses( sqlFilter, inClause );
+    }
+  }
+
+  //IMPORTANT - this MUST be the last clause added
```

Comment:
Can you add the reason why this has to be the last clause?

Reply:
It is a copy-paste leftover from another provider. I cannot tell you why this must be the last clause.